File: solutions/19. Remove Nth Node From End of List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        // if (head.next == null) return null;
        // if (head.next.next == null) return head.next;
​
        ListNode slow = head;
        ListNode fast = head;
​
        for (int i = 0; i < n; i++) {
            fast = fast.next;
        }
​
        if (fast == null) return slow.next;
​
        while (fast.next != null) {
            slow = slow.next;
            fast = fast.next;
        }
​
        slow.next = slow.next.next;
        
        return head;
    }
}

File: ..solutions/19. Remove Nth Node From End of List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int k) {
        ListNode* root = new ListNode() ;
        root->next = head;
        ListNode* f = root, *s = root;
        for(int i = 1; i <= k+1; i++)
            f = f -> next;
        
        while(f){
            f = f -> next;
            s = s -> next;
        }
        ListNode* d = s->next;
        s->next = s->next->next;
        delete d;
        return root->next;
    }
};
